File: day10/part1/c/main.c
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE 4096
#define MAX_LIGHTS 16
#define MAX_BUTTONS 20

int parse_lights(const char *start, int *num_lights) {
  int target = 0;
  *num_lights = 0;
  const char *p = start + 1;
  while (*p && *p != ']') {
    if (*p == '#') {
      target |= (1 << *num_lights);
    }
    (*num_lights)++;
    p++;
  }
  return target;
}

int parse_button(const char *start, const char **end) {
  int mask = 0;
  const char *p = start + 1;
  while (*p && *p != ')') {
    if (*p >= '0' && *p <= '9') {
      int num = 0;
      while (*p >= '0' && *p <= '9') {
        num = num * 10 + (*p - '0');
        p++;
      }
      mask |= (1 << num);
    } else {
      p++;
    }
  }
  if (*p == ')')
    p++;
  *end = p;
  return mask;
}

int min_presses(int target, int *buttons, int num_buttons, int num_lights) {

  int max_state = 1 << num_lights;
  int *dist = malloc(max_state * sizeof(int));
  for (int i = 0; i < max_state; i++)
    dist[i] = INT_MAX;

  int *queue = malloc(max_state * sizeof(int));
  int front = 0, back = 0;

  dist[0] = 0;
  queue[back++] = 0;

  while (front < back) {
    int state = queue[front++];
    if (state == target) {
      int result = dist[target];
      free(dist);
      free(queue);
      return result;
    }

    for (int i = 0; i < num_buttons; i++) {
      int next_state = state ^ buttons[i];
      if (dist[next_state] == INT_MAX) {
        dist[next_state] = dist[state] + 1;
        queue[back++] = next_state;
      }
    }
  }

  int result = dist[target];
  free(dist);
  free(queue);
  return result == INT_MAX ? -1 : result;
}

int main(void) {
  FILE *fp = fopen("../list.txt", "r");
  if (!fp) {
    perror("Failed to open list.txt");
    return 1;
  }

  char line[MAX_LINE];
  long long total = 0;

  while (fgets(line, sizeof(line), fp)) {

    if (line[0] == '\n' || line[0] == '\0')
      continue;

    char *bracket = strchr(line, '[');
    if (!bracket)
      continue;

    int num_lights;
    int target = parse_lights(bracket, &num_lights);

    int buttons[MAX_BUTTONS];
    int num_buttons = 0;

    const char *p = strchr(bracket, ']');
    if (!p)
      continue;
    p++;

    while (*p) {

      while (*p && *p != '(' && *p != '{')
        p++;
      if (*p == '{' || *p == '\0')
        break;

      const char *end;
      buttons[num_buttons++] = parse_button(p, &end);
      p = end;
    }

    int min = min_presses(target, buttons, num_buttons, num_lights);
    if (min >= 0) {
      total += min;
    }
  }

  fclose(fp);
  printf("%lld\n", total);

  return 0;
}

File: day10/part1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strings"
)

func parseLights(s string) (target int, numLights int) {
	target = 0
	numLights = 0
	for i := 1; i < len(s) && s[i] != ']'; i++ {
		if s[i] == '#' {
			target |= 1 << numLights
		}
		numLights++
	}
	return target, numLights
}

func parseButton(s string) (mask int, end int) {
	mask = 0
	i := 1
	for i < len(s) && s[i] != ')' {
		if s[i] >= '0' && s[i] <= '9' {
			num := 0
			for i < len(s) && s[i] >= '0' && s[i] <= '9' {
				num = num*10 + int(s[i]-'0')
				i++
			}
			mask |= 1 << num
		} else {
			i++
		}
	}
	if i < len(s) && s[i] == ')' {
		i++
	}
	return mask, i
}

func minPresses(target int, buttons []int, numLights int) int {
	maxState := 1 << numLights
	dist := make([]int, maxState)
	for i := range dist {
		dist[i] = math.MaxInt32
	}

	queue := make([]int, 0, maxState)
	dist[0] = 0
	queue = append(queue, 0)

	for front := 0; front < len(queue); front++ {
		state := queue[front]
		if state == target {
			return dist[target]
		}

		for _, btn := range buttons {
			nextState := state ^ btn
			if dist[nextState] == math.MaxInt32 {
				dist[nextState] = dist[state] + 1
				queue = append(queue, nextState)
			}
		}
	}

	if dist[target] == math.MaxInt32 {
		return -1
	}
	return dist[target]
}

func main() {
	file, err := os.Open("../list.txt")
	if err != nil {
		fmt.Fprintln(os.Stderr, "Failed to open list.txt:", err)
		os.Exit(1)
	}
	defer file.Close()

	var total int64 = 0
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()

		if line == "" {
			continue
		}

		bracketIdx := strings.Index(line, "[")
		if bracketIdx == -1 {
			continue
		}

		target, numLights := parseLights(line[bracketIdx:])

		var buttons []int

		closeBracketIdx := strings.Index(line[bracketIdx:], "]")
		if closeBracketIdx == -1 {
			continue
		}
		p := bracketIdx + closeBracketIdx + 1

		for p < len(line) {

			for p < len(line) && line[p] != '(' && line[p] != '{' {
				p++
			}
			if p >= len(line) || line[p] == '{' {
				break
			}

			mask, end := parseButton(line[p:])
			buttons = append(buttons, mask)
			p += end
		}

		minVal := minPresses(target, buttons, numLights)
		if minVal >= 0 {
			total += int64(minVal)
		}
	}

	fmt.Println(total)
}
